File: lab11/w3.cpp
#include <iostream>
#include <queue>
#include <string>

class Queue {
private:
    std::queue<std::string> items;

public:
    bool is_empty() const {
        return items.empty();
    }

    void enqueue(const std::string& item) {
        items.push(item);
    }

    void dequeue() {
        if (!is_empty()) {
            items.pop();
        } else {
            throw std::out_of_range("dequeue from empty queue");
        }
    }

    size_t size() const {
        return items.size();
    }

    std::string front() const {
        if (!is_empty()) {
            return items.front();
        } else {
            throw std::out_of_range("front from empty queue");
        }
    }
};

// Real-world example: Ticketing system
int main() {
    Queue ticket_queue;

    // Customers arriving at the ticket counter
    ticket_queue.enqueue("Customer 1");
    ticket_queue.enqueue("Customer 2");
    ticket_queue.enqueue("Customer 3");

    std::cout << "Queue size: " << ticket_queue.size() << std::endl;

    // Serving customers
    std::cout << "Serving: " << ticket_queue.front() << std::endl;
    ticket_queue.dequeue();
    std::cout << "Serving: " << ticket_queue.front() << std::endl;
    ticket_queue.dequeue();
    std::cout << "Queue size: " << ticket_queue.size() << std::endl;

    return 0;
}

File: lab_11/w3.py
class Queue:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return len(self.items) == 0

    def enqueue(self, item):
        self.items.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)
        else:
            raise IndexError("dequeue from empty queue")

    def size(self):
        return len(self.items)

# Real-world example: Ticketing system
ticket_queue = Queue()

# Customers arriving at the ticket counter
ticket_queue.enqueue("Customer 1")
ticket_queue.enqueue("Customer 2")
ticket_queue.enqueue("Customer 3")

print("Queue size:", ticket_queue.size())

# Serving customers
print("Serving:", ticket_queue.dequeue())
print("Serving:", ticket_queue.dequeue())
print("Queue size:", ticket_queue.size())
